### PR TITLE
LOOP-2820 notification settings interface

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-2820

### CODE REVIEW

File: DashKit/DashPumpManager.swift [hunk 1]
```diff
@@ -376,6 +376,13 @@ open class DashPumpManager: PumpManager {
                     state.updateFromPodStatus(status: status)
                 })
             }
+            if case .event(let event) = activationStatus, case .step1Completed = event {
+                if let activationDate = self.state.podActivatedAt {
+                    self.mutateState { (state) in
+                        state.expirationReminderDate = activationDate + .days(3) - podExpirationAlert.intervalBeforeExpiration
```
Comment (nhamming) [inline DashKit/DashPumpManager.swift]: Should `.days(3)` => `Pod.lifetime`?

File: DashKitUI/View Models/DashSettingsViewModel.swift [hunk 2]
```diff
@@ -112,6 +135,9 @@ class DashSettingsViewModel: ObservableObject {
         basalDeliveryState = pumpManager.status.basalDeliveryState
         basalDeliveryRate = self.pumpManager.basalDeliveryRate
         reservoirLevel = self.pumpManager.reservoirLevel
+        expirationReminderDate = self.pumpManager.state.expirationReminderDate
+        expirationReminderDefault = Int(self.pumpManager.defaultExpirationReminderOffset.hours)
```
Comment (nhamming) [inline DashKitUI/View Models/DashSettingsViewModel.swift:139]: [nit] include `Offset` in the name?

Also, curious why this comes from the `pumpManager` when the others are pulled from `state`? Should they be consistent?

**Review by nhamming — APPROVED** [review]
LGTM

File: DashKitUI/View Models/DashSettingsViewModel.swift (hunk 2 above)
Comment (ps2) [reply DashKitUI/View Models/DashSettingsViewModel.swift:139]: Most are coming from pump manager as they should.  Having the UI reaching into state is a little smelly, imo, but I hate making getters/accessors (more code) that are just boilerplate code.  If you think it's worth making getters/setters for these on pumpManager, lmk.

File: DashKitUI/Views/DesignElements/RoundedCard.swift [hunk 3]
```diff
@@ -0,0 +1,130 @@
+//
+//  RoundedCard.swift
+//  SectionedUIPlaygroundApp
```
Comment (darinkrauss) [inline DashKitUI/Views/DesignElements/RoundedCard.swift]: [nit] DashKitUI?

File: DashKitUI/Views/DesignElements/RoundedCard.swift [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    var text: String
+    
+    init(_ text: String) {
+        self.text = text
+    }
+    
+    var body: some View {
+        Text(text)
+            .font(.caption)
+            .fixedSize(horizontal: false, vertical: /*@START_MENU_TOKEN@*/true/*@END_MENU_TOKEN@*/)
```
Comment (darinkrauss) [inline DashKitUI/Views/DesignElements/RoundedCard.swift]: [nit] What is `/*@START_MENU_TOKEN@*/true/*@END_MENU_TOKEN@*/`?

File: DashKitUI/Views/LowReservoirReminderEditView.swift [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        VStack {
+            RoundedCardScrollView(title: LocalizedString("Low Reservoir Reminder", comment: "Title for low reservoir reminder edit page")) {
+                RoundedCard {
+                    RoundedCardValueRow(
+                        label: LocalizedString("Low Reservoir Reminder", comment: "Label for low reservoir reminder row"),
+                        value: formatValue(selectedValue),
+                        highlightValue: true
+                    )
+                    Picker("", selection: $selectedValue) {
+                        ForEach(10..<50, id: \.self) { value in
```
Comment (darinkrauss) [inline DashKitUI/Views/LowReservoirReminderEditView.swift]: [nit] Should these be called out as constants somewhere? Seems hidden here.

File: DashKitUI/Views/NotificationSettingsView.swift [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                footer: "The App notifies you in advance of Pod expiration.  Set the number of hours advance notice you would like to have."
+            ) {
+                expirationReminderRow
+                if let scheduledReminderDate = scheduledReminderDate  {
+                    Divider()
+                    scheduledReminderRow(scheduledDate: scheduledReminderDate)
+                }
+            }
+
+            RoundedCard(footer: "The App notifies you when the amount of insulin in the Pod reaches this level.") {
```
Comment (darinkrauss) [inline DashKitUI/Views/NotificationSettingsView.swift]: [nit] "The App"?

File: DashKitUI/Views/NotificationSettingsView.swift [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    var onSaveScheduledExpirationReminder: ((_ selectedDate: Date, _ completion: @escaping (_ error: Error?) -> Void) -> Void)?
+    
+    var onSaveLowReservoirReminder: ((_ selectedValue: Int, _ completion: @escaping (_ error: Error?) -> Void) -> Void)?
+    
+    var insulinQuantityFormatter = QuantityFormatter(for: .internationalUnit())
+
+    var body: some View {
+        RoundedCardScrollView {
+            RoundedCard(
+                title: "Omnipod Reminders",
```
Comment (darinkrauss) [inline DashKitUI/Views/NotificationSettingsView.swift]: [nit] Localize all of these new strings?

File: DashKitUI/Views/NotificationSettingsView.swift [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                Spacer()
+                Button("\(expirationReminderDefault) h") {
+                    withAnimation {
+                        showingHourPicker.toggle()
+                    }
+                }
+            }
+            if showingHourPicker {
+                Picker("Expiration Reminder Default", selection: $expirationReminderDefault) {
+                    ForEach(1..<30, id: \.self) { value in
```
Comment (darinkrauss) [inline DashKitUI/Views/NotificationSettingsView.swift]: [nit] Constants?

**Review by darinkrauss — APPROVED** [review]
A number of nits, but otherwise LGTM! 🚀

File: DashKitUI/Views/DesignElements/RoundedCard.swift (hunk 4 above)
Comment (ps2) [reply DashKitUI/Views/DesignElements/RoundedCard.swift]: Ah, that's xcode placeholder text. Will remove.

File: DashKitUI/Views/NotificationSettingsView.swift (hunk 6 above)
Comment (ps2) [reply DashKitUI/Views/NotificationSettingsView.swift]: That's how it is in the design. I think it's fine; but would be easy to change to app name if we want.

Comment (darinkrauss) [reply DashKitUI/Views/NotificationSettingsView.swift]: Nope, not necessary.

**Pull request merged**